File: problems/387_first_unique_character_in_string.py
'''
Given a string, find the first non-repeating character in it and return it's index. If it doesn't exist, return -1.

Examples:

s = "leetcode"
return 0.

s = "loveleetcode",
return 2.
Note: You may assume the string contain only lowercase letters.
'''

from collections import defaultdict

class Solution:
    def firstUniqChar(self, s: str) -> int:
        count_by_char = defaultdict(int)
        for char in s:
            count_by_char[char] += 1
        for i, char in enumerate(s):
            if count_by_char[char] == 1:
                return i
        return -1

File: problems/387_first_unique_character_in_string.go
/*
Given a string s, find the first non-repeating character in it and return its index. If it does not exist, return -1.

 

Example 1:

Input: s = "leetcode"
Output: 0
Example 2:

Input: s = "loveleetcode"
Output: 2
Example 3:

Input: s = "aabb"
Output: -1
 

Constraints:

1 <= s.length <= 105
s consists of only lowercase English letters.
*/

func firstUniqChar(s string) int {
    counts := make([]int, 26)
    for _, v := range s {
        counts[v - 'a']++
    }
    for i, v := range s {
        if counts[v - 'a'] == 1 {
            return i
        }
    }

    return -1
}
